src/wavm.cpp

```diff
@@ -0,0 +1,194 @@
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "codeCopy", void, codeCopy, U32 resultOffset, U32 codeOffset, U32 length)
+  {
+    interface.top()->eeiCodeCopy(resultOffset, codeOffset, length);
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "getCodeSize", U32, getCodeSize)
+  {
+    return interface.top()->eeiGetCodeSize();
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "finish", void, finish, U32 dataOffset, U32 length)
+  {
+    interface.top()->eeiFinish(dataOffset, length);
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "revert", void, revert, U32 dataOffset, U32 length)
+  {
+    interface.top()->eeiRevert(dataOffset, length);
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "getReturnDataSize", U32, getReturnDataSize)
+  {
+    return interface.top()->eeiGetReturnDataSize();
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "returnDataCopy", void, returnDataCopy, U32 resultOffset, U32 dataOffset, U32 length)
+  {
+    return interface.top()->eeiReturnDataCopy(resultOffset, dataOffset, length);
+  }
+
+
+  // this is needed for resolving names of imported host functions
+  struct HeraWavmResolver : Runtime::Resolver {
+    Runtime::Compartment* compartment;
+    HashMap<std::string, Runtime::ModuleInstance*> moduleNameToInstanceMap;
+
+    HeraWavmResolver(Runtime::Compartment* inCompartment) : compartment(inCompartment) {}
+
+    bool resolve(const std::string& moduleName,
+           const std::string& exportName,
+           IR::ObjectType type,
+           Runtime::Object*& outObject) override {
+      auto namedInstance = moduleNameToInstanceMap.get(moduleName);
+      outObject = getInstanceExport(*namedInstance, exportName);
+      return true;
+    }
+  };
+
+} //namespace wavm_host_module
+
+
+
+ExecutionResult WavmEngine::execute(
+  evmc_context* context,
+  vector<uint8_t> const& code,
+  vector<uint8_t> const& state_code,
+  evmc_message const& msg,
+  bool meterInterfaceGas
+) {
+  HERA_DEBUG << "Executing with wavm...\n";
+
+  // set up a new ethereum interface just for this contract invokation
+  ExecutionResult result;
+  WavmEthereumInterface myInterface{context, state_code, msg, result, meterInterfaceGas};
+  wavm_host_module::interface.push(&myInterface);
+
+  // first parse module
+  std::vector<U8> codeBytes = static_cast< std::vector<U8> > (code);
+  IR::Module moduleAST; 
+  bool loadedSuccess = loadBinaryModule(codeBytes.data(), codeBytes.size(), moduleAST);
+  heraAssert(loadedSuccess, "wavm couldn't parse module into syntax tree");
+
+  // next set up the host module. Note: in ewasm, we create a new VM for each call to a module, so we must instantiate a new host module for each of these VMs, this is inefficient, but OK for prototyping.
+  // compartment is like the Wasm store, represents the VM, has lists of globals, memories, tables, and also has wavm's runtime stuff
+  Runtime::GCPointer<Runtime::Compartment> compartment = Runtime::createCompartment();
```

Comment:
Not fully convinced we should use `Runtime::GCPointer` and not `std::unique_ptr` if because of that we need to call `Runtime::collectGarbage()`. However if we need to call that anyway, then I am fine.

Reply:
WAVM uses some custom abstracted constructor-like things, and I don't think that they can be cleaned up through `std::unique_ptr`.

Reply:
@chfast can you take a look?

Reply:
When you use it and it does not crash you are good.